net/ieee802154: add PHY modes caps for HAL

jeandudey opened the pull request:
### Contribution description

Adds a new method/callback on `ieee802154_radio_ops_t` structure, to get the supported PHY modes on the devices.

The motivation is to get the 2047 PDU (minus overhead bytes) on Multi-Regional PHYs, found on IEEE 802.15.4g transceivers, such as the at86rf215, or the cc13x2 MCUs and to calculate the timings on runtime for the `ACK_TIMEOUT` constant for the different PHYs that exist.

I slightly modified the PHY modes enum so, PHY modes can be grouped together when PHY capabilities are being checked. 

The idea is that on `config_phy` the devices select the PHY mode, and configure the device for it, for example, the at86rf215 device implements various PHY modes, this method would allow to specify one of all the PHYs it supports. For other devices that only implement only one PHY mode doesn't make too much sense though, maybe when PHY specific parameters are added to `ieee802154_phy_conf_t` it could have an use case.

### Testing procedure

- `make -C examples/gnrc_networking BOARD=cc2538dk` or on any supported board that uses the HAL, `ifconfig` should output the PHY mode.

### Issues/PRs references

- Related to #13295

benpicco commented on sys/net/link_layer/ieee802154/submac.c:
```diff
@@ -337,6 +337,25 @@ int ieee802154_submac_init(ieee802154_submac_t *submac, const network_uint16_t *
+         * modes definition, first IEEE 802.15.4-2006 PHY modes, then
+         * IEEE 802.15.4g-2012 PHY modes. */
+        uint8_t bit;
+        bitarithm_test_and_clear(phy_modes, &bit);
```
I think you rather want to use `bitarithm_lsb()` here - the order in which bits are processed by `bitarithm_test_and_clear()` is not defined.

jeandudey replied:
Thanks, solved at ea5ea1913a

jia200x commented:
In general, I think this goes in the same direction as the original thoughts regarding how to treat the PHY modes :)

> The idea is that on config_phy the devices select the PHY mode, and configure the device for it, for example, the at86rf215 device implements various PHY modes, this method would allow to specify one of all the PHYs it supports. For other devices that only implement only one PHY mode doesn't make too much sense though, maybe when PHY specific parameters are added to ieee802154_phy_conf_t it could have an use case.

This is exactly what was discussed in #13943. The reason why we defined an `ieee802154_phy_conf_t` was to allow extending the HAL to the IEEE 802.15.4g radios. So this sounds quite reasonable.

> Maybe another approach could be taken using caps (IEEE802154_CAP_MR_FSK, IEEE802154_CAP_MR_OQPSK, and such) to implement this functionality when the SubMAC initializes the device. That said, I slightly modified the PHY modes enum so in one call all PHY modes could be retrieved.

I think I tend to prefer this approach. There were even some comments about removing the "get_caps" pointer and simply store a bitmap to indicate all caps. This makes it easier to optimize out code beforehand when using only one radio (e.g resolve caps function to `true` or `false` at compile time and save some memory).

jia200x commented:
>  For other devices that only implement only one PHY mode doesn't make too much sense though, maybe when PHY specific parameters are added to ieee802154_phy_conf_t it could have an use case.

This stuff can be solved with compile time optimizations (e.g don't add the struct member if it's not needed)

jia200x commented on sys/net/link_layer/ieee802154/submac.c:
```diff
@@ -337,6 +337,43 @@ int ieee802154_submac_init(ieee802154_submac_t *submac, const network_uint16_t *
 
+    /* Get supported PHY modes */
+    int phy_modes = 0;
+    if (ieee802154_radio_has_phy_bpsk(dev)) {
```
if we change the `get_caps` to a flag based solution, we could allocate the bitmask beforehand an simply do e.g:

```c
phy_modes = dev->caps & IEEE802154_PHY_MODES_MASK;
```

jia200x replied:
but I would keep these ieee802154_radio_has_phy_xxx, since those can be handy anyway :).

jeandudey replied:
>if we change the get_caps to a flag based solution, we could allocate the bitmask beforehand an simply do e.g:

Should it be provided as a separate PR then? Or could I modify it directly here on this PR.

jia200x replied:
I think in any case in a different PR. Since the Radio HAL [is still under discussion](https://github.com/RIOT-OS/RIOT/pull/13943), I think it's worth to comment it in the RDM.

benpicco commented:
Looks like this could need a rebase

jeandudey commented:
Rebased and squashed the commits.

I removed the modification to PHY modes enumeration and instead used the IEEE 802.15.4 capabilities one for that purpose, added the respective conversion function between PHY mode <-> capability.

And changed the `uint16_t` to an `uint32_t` of the capabilities to hold the PHY mode bits, the layout size stays the same because of the aligment.

Modified the HAL test to check for PHY modes and the PHY configuration command to specify the mode.

benpicco commented:
Looks good to me, what do you think @jia200x?

jia200x commented:
I will give it a final look again tomorrow morning, but from what I see it looks neat 👌

jia200x commented:
it looks nice!

jia200x commented:
I get this when compiling `examples/gnrc_networking` with `remote-revb`:
```
/home/jialamos/Development/RIOT/sys/include/net/ieee802154.h:214:14: error: missing binary operator before token "("
 #if IS_ACTIVE(CONFIG_IEEE802154_DEFAULT_PHY_BPSK)
              ^
/home/jialamos/Development/RIOT/sys/include/net/ieee802154.h:216:16: error: missing binary operator before token "("
 #elif IS_ACTIVE(CONFIG_IEEE802154_DEFAULT_PHY_ASK)
                ^
/home/jialamos/Development/RIOT/sys/include/net/ieee802154.h:218:16: error: missing binary operator before token "("
 #elif IS_ACTIVE(CONFIG_IEEE802154_DEFAULT_PHY_OQPSK)
                ^
/home/jialamos/Development/RIOT/sys/include/net/ieee802154.h:220:16: error: missing binary operator before token "("
 #elif IS_ACTIVE(CONFIG_IEEE802154_DEFAULT_PHY_MR_OQPSK)
                ^
/home/jialamos/Development/RIOT/sys/include/net/ieee802154.h:222:16: error: missing binary operator before token "("
 #elif IS_ACTIVE(CONFIG_IEEE802154_DEFAULT_PHY_MR_OFDM)
                ^
/home/jialamos/Development/RIOT/sys/include/net/ieee802154.h:224:16: error: missing binary operator before token "("
 #elif IS_ACTIVE(CONFIG_IEEE802154_DEFAULT_PHY_MR_FSK)
                ^
```

There's a missing include to `kernel_defines.h`. Feel free to add this patch:
```
diff --git a/sys/include/net/ieee802154.h b/sys/include/net/ieee802154.h
index d4c818f665..17769ea4c7 100644
--- a/sys/include/net/ieee802154.h
+++ b/sys/include/net/ieee802154.h
@@ -26,6 +26,7 @@
 
 #include "byteorder.h"
 #include "net/eui64.h"
+#include "kernel_defines.h"
 
 #ifdef __cplusplus
 extern "C" {

```

jeandudey commented:
@jia200x may I rebase and & squash directly the patch?

jia200x commented:
Otherwise, it works:

Remote-revb:
```
2021-02-26 11:10:19,408 # main(): This is RIOT! (Version: 2021.04-devel-690-gb6553-2021_02_04-ieee802154-phy)
2021-02-26 11:10:19,411 # Initialization successful - starting the shell now
help
2021-02-26 11:10:21,489 # help
2021-02-26 11:10:21,490 # Command              Description
2021-02-26 11:10:21,492 # ---------------------------------------
2021-02-26 11:10:21,495 # config_phy           Set channel and TX power
2021-02-26 11:10:21,499 # print_addr           Print IEEE802.15.4 addresses
2021-02-26 11:10:21,506 # txtsnd               Send IEEE 802.15.4 packet
2021-02-26 11:10:21,508 # test_states          Test state changes
2021-02-26 11:10:21,509 # cca                  Perform CCA
2021-02-26 11:10:21,513 # config_cca           Config CCA parameters
2021-02-26 11:10:21,522 # rx_mode              Enable/Disable AACK or set Frame Pending bit or set promiscuos mode
2021-02-26 11:10:21,533 # tx_mode              Enable CSMA-CA, CCA or direct transmission
2021-02-26 11:10:21,535 # caps                 Get a list of caps supported by the device
caps
2021-02-26 11:10:23,889 # caps
2021-02-26 11:10:23,890 #                CAPS              
2021-02-26 11:10:23,891 # =================================
2021-02-26 11:10:23,893 # - Frame Retransmissions:        n
2021-02-26 11:10:23,894 #     * Info retries:             n
2021-02-26 11:10:23,895 # - Auto CSMA-CA:                 y
2021-02-26 11:10:23,897 # - 2.4 GHz band:                 y
2021-02-26 11:10:23,898 # - SubGHz band:                  n
2021-02-26 11:10:23,912 # - TX DONE indication:           y
2021-02-26 11:10:23,914 #     * ACK Timeout indication:   n
2021-02-26 11:10:23,916 # - RX_START indication:          y
2021-02-26 11:10:23,917 # - CCA Done indication:          y
2021-02-26 11:10:23,918 # - PHY modes:                      
2021-02-26 11:10:23,920 # -   * BPSK:                     n
2021-02-26 11:10:23,921 # -   * ASK:                      n
2021-02-26 11:10:23,922 # -   * O-QPSK:                   y
2021-02-26 11:10:23,924 # -   * MR-O-QPSK:                n
2021-02-26 11:10:23,925 # -   * MR-OFDM:                  n
2021-02-26 11:10:23,926 # -   * MR-FSK:                   n
```

```
/home/jialamos/Development/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-02-26 11:13:24,481 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
caps
2021-02-26 11:13:34,868 # caps
2021-02-26 11:13:34,871 #                CAPS              
2021-02-26 11:13:34,876 # =================================
2021-02-26 11:13:34,878 # - Frame Retransmissions:        n
2021-02-26 11:13:34,880 #     * Info retries:             n
2021-02-26 11:13:34,883 # - Auto CSMA-CA:                 n
2021-02-26 11:13:34,886 # - 2.4 GHz band:                 y
2021-02-26 11:13:34,889 # - SubGHz band:                  n
2021-02-26 11:13:34,892 # - TX DONE indication:           y
2021-02-26 11:13:34,895 #     * ACK Timeout indication:   n
2021-02-26 11:13:34,898 # - RX_START indication:          y
2021-02-26 11:13:34,901 # - CCA Done indication:          y
2021-02-26 11:13:34,904 # - PHY modes:                      
2021-02-26 11:13:34,907 # -   * BPSK:                     n
2021-02-26 11:13:34,910 # -   * ASK:                      n
2021-02-26 11:13:34,913 # -   * O-QPSK:                   y
2021-02-26 11:13:34,916 # -   * MR-O-QPSK:                n
2021-02-26 11:13:34,919 # -   * MR-OFDM:                  n
2021-02-26 11:13:34,922 # -   * MR-FSK:                   n
> 
```

jia200x commented:
> @jia200x may I rebase and & squash directly the patch?

Sure!

jeandudey commented:
Done!

jia200x reviewed (approved):
ACK
